// @stackra/tsup-config@1.0.6/dist/index.js
import { resolve } from 'path';
import { readFileSync } from 'fs';
import { build } from 'tsup';

// src/utils/build-banner.ts
function buildBanner(pkg) {
  const name = pkg.name || "Unknown Package";
  const version = pkg.version || "0.0.0";
  const author = pkg.author || "Unknown Author";
  const license = pkg.license || "UNLICENSED";
  return `/**
 * ${name} v${version}
 * (c) ${(/* @__PURE__ */ new Date()).getFullYear()} ${author}
 * @license ${license}
 */`;
}
function loadPackageJson() {
  const pkgPath = resolve(process.cwd(), "package.json");
  try {
    const content = readFileSync(pkgPath, "utf-8");
    return JSON.parse(content);
  } catch (error) {
    throw new Error(
      `Failed to load package.json from ${pkgPath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// src/utils/compute-externals.ts
function computeExternals(pkg) {
  const deps = Object.keys(pkg.dependencies || {});
  const devDeps = Object.keys(pkg.devDependencies || {});
  const peerDeps = Object.keys(pkg.peerDependencies || {});
  return [.../* @__PURE__ */ new Set([...deps, ...devDeps, ...peerDeps])];
}

// src/presets/base.ts
var preset = (() => {
  const pkg = loadPackageJson();
  const externals = computeExternals(pkg);
  const banner = buildBanner(pkg);
  return {
    // Entry point(s) for the build
    entry: ["src/index.ts"],
    // Output formats: ES Module and CommonJS
    format: ["esm", "cjs"],
    // Generate TypeScript declaration files (.d.ts)
    dts: true,
    // Clean output directory before build
    clean: true,
    // Disable code splitting (simpler output)
    splitting: false,
    // Enable tree-shaking to remove unused code
    treeshake: true,
    // Mark all dependencies as external (won't be bundled)
    external: ["figlet", ...externals],
    // Add license banner to JavaScript output
    banner: {
      js: banner
    },
    // Generate source maps in development only
    sourcemap: process.env.NODE_ENV === "development",
    // Don't minify output (keeps code readable)
    minify: false,
    // Target ES2022 for modern JavaScript features
    target: "es2022",
    // Output directory
    outDir: "dist",
    // ✅ Always bundle by default, so consumers don't have to override
    bundle: true
  };
})();

// src/presets/cli-lib.ts
var preset2 = (() => {
  const pkg = loadPackageJson();
  const externals = computeExternals(pkg);
  const banner = buildBanner(pkg);
  return {
    entry: ["src/index.ts"],
    // CJS only (Node.js standard)
    format: ["cjs"],
    // No type declarations for CLI tools
    dts: false,
    // Clean before build
    clean: true,
    // Single file output
    splitting: false,
    // Remove unused code
    treeshake: true,
    // External dependencies
    external: externals,
    // Shebang + license banner
    banner: {
      js: `#!/usr/bin/env node
${banner}`
    },
    // Source maps in dev only
    sourcemap: process.env.NODE_ENV === "development",
    // Keep readable
    minify: false,
    // Node18 LTS target
    target: "node18",
    outDir: "dist",
    // Bundle everything
    bundle: true
  };
})();

// src/presets/nest-lib.ts
var preset3 = {
  // Inherit base configuration (ESM + CJS, types, etc.)
  ...preset,
  // Don't bundle - preserve module structure for proper tree-shaking
  // This allows consumers to only import what they need
  bundle: false,
  // No code splitting needed for libraries
  splitting: false,
  // Use project's tsconfig.json for proper decorator handling
  // NestJS requires "emitDecoratorMetadata": true
  tsconfig: "./tsconfig.json"
};

// src/presets/nest-app.ts
var preset4 = (() => {
  const pkg = loadPackageJson();
  const externals = computeExternals(pkg);
  const banner = buildBanner(pkg);
  return {
    // NestJS application entry point (convention)
    entry: ["src/main.ts"],
    // CommonJS format only (NestJS runtime requirement)
    format: ["cjs"],
    // No type declarations for applications
    dts: false,
    // Clean output directory before build
    clean: true,
    // Single file output (no code splitting)
    splitting: false,
    // Remove unused code
    treeshake: true,
    // External dependencies (not bundled)
    external: externals,
    // Shebang for executability + license banner
    banner: {
      js: `#!/usr/bin/env node
${banner}`
    },
    // Source maps in development only
    sourcemap: process.env.NODE_ENV === "development",
    // Keep code readable
    minify: false,
    // Node.js 18 LTS target
    target: "node18",
    // Output directory
    outDir: "dist",
    // Bundle all dependencies into single file
    bundle: true,
    // Use project's TypeScript config (for decorators/metadata)
    tsconfig: "./tsconfig.json"
  };
})();

// src/presets/react-lib.ts
var preset5 = {
  // Inherit base configuration
  ...preset,
  // Mark React as external (consumer provides it)
  // Prevents bundling React multiple times in consumer apps
  external: [
    ...Array.isArray(preset.external) ? preset.external : [],
    "react",
    // Core React library
    "react-dom",
    // React DOM renderer
    "react/jsx-runtime"
    // New JSX transform runtime
  ],
  // Configure esbuild for modern JSX
  esbuildOptions(options) {
    options.jsx = "automatic";
  }
};
var tsupPlugin = (options) => {
  return {
    /**
     * Plugin identifier used by Vite for debugging and plugin ordering.
     * This name appears in build logs and can be used to filter plugins.
     */
    name: "vite-plugin-tsup",
    /**
     * Specifies when this plugin should be active.
     *
     * Setting `apply: 'build'` ensures the plugin only runs during production builds,
     * not during development server execution. This is important because:
     * - tsup bundling is typically only needed for production artifacts
     * - Running tsup during dev would slow down HMR and development experience
     * - Vite's dev server already handles module transformation efficiently
     */
    apply: "build",
    /**
     * Vite lifecycle hook that runs after the config has been fully resolved.
     *
     * This hook is called once Vite has processed all configuration (including
     * user config, environment variables, and plugin-applied modifications).
     * At this point, we have access to the complete build context.
     *
     * **Why use configResolved instead of buildStart?**
     * - configResolved runs before any build operations begin
     * - It ensures tsup completes before Vite starts its own bundling
     * - This prevents race conditions or file conflicts
     *
     * **Error Handling:**
     * If tsup build fails, the error will propagate and fail the entire Vite build,
     * which is the desired behavior to prevent shipping broken builds.
     *
     * @param _config - The resolved Vite configuration (unused but available)
     *                  Prefixed with underscore to indicate intentionally unused parameter
     *
     * @throws Will throw if tsup build fails (compilation errors, file system issues, etc.)
     */
    async configResolved(_config) {
      await build(options);
    }
  };
};

export { preset as basePreset, buildBanner, preset2 as cliLibPreset, computeExternals, loadPackageJson, preset4 as nestAppPreset, preset3 as nestLibPreset, preset5 as reactLibPreset, tsupPlugin };
